### PR TITLE
BUG: DataFrame.__setitem__ not raising ValueError when rhs is df and has wrong number of columns

### LINKED ISSUE #38604: BUG: surprising non-error when assigning a multi-column DataFrame to a single column
- [x] I have checked that this issue has not already been reported.

- [x] I have confirmed this bug exists on the latest version of pandas.

- [x] (optional) I have confirmed this bug exists on the master branch of pandas.

---

Assigning a multi-column DataFrame to a single column does not raise an error and simply assigns the first column of the DataFrame to the column. Below is an example:

```python
df1 = pd.DataFrame({'a': [0, 1, 2, 3], 'b': [4, 5, 6, 7]}) 
df2 = pd.DataFrame({'c': [8, 9, 10, 11], 'd': [12, 13, 14, 15]}) 
df1['b'] = df2
```

This seems like one of the instances where not raising an error might ease some short-term pain but could mask underlying issues with a user's code. This behavior seems to have been around for a while. It occurs in master and as far back as 1.0.3.

### PR DESCRIPTION
**Pull request by phofl**
- [x] closes #38604
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [x] whatsnew entry

Have to check how many columns key references before raising (duplicates or MultiIndex may reference more than one column)

### CODE REVIEW

File: pandas/core/frame.py [hunk 1]
```diff
@@ -3256,14 +3256,20 @@ def _setitem_frame(self, key, value):
     def _set_item_frame_value(self, key, value: DataFrame) -> None:
         self._ensure_valid_index(value)
 
-        # align right-hand-side columns if self.columns
-        # is multi-index and self[key] is a sub-frame
-        if isinstance(self.columns, MultiIndex) and key in self.columns:
+        if key in self.columns:
             loc = self.columns.get_loc(key)
```
Comment (jreback) [inline pandas/core/frame.py:3260]: can you add a comment: `# align columns ` before L3259

File: pandas/core/frame.py [hunk 2]
```diff
@@ -3256,14 +3256,20 @@ def _setitem_frame(self, key, value):
...
-                    value = value.reindex(cols, axis=1)
+            cols = self.columns[loc]
+            len_cols = 1 if is_scalar(cols) else len(cols)
+            if len_cols != len(value.columns):
+                raise ValueError("Columns must be same length as key")
+
+            # align right-hand-side columns if self.columns
+            # is multi-index and self[key] is a sub-frame
+            if isinstance(self.columns, MultiIndex):
+                if isinstance(loc, (slice, Series, np.ndarray, Index)):
```
Comment (jreback) [inline pandas/core/frame.py:3269]: can combine these conditions

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/frame.py (hunk 1 above)
Comment (phofl) [reply pandas/core/frame.py:3260]: Done

File: pandas/core/frame.py (hunk 2 above)
Comment (phofl) [reply pandas/core/frame.py:3269]: Done

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @phofl

File: pandas/tests/frame/indexing/test_setitem.py [hunk 3]
```diff
@@ -366,6 +366,23 @@ def test_setitem_frame_duplicate_columns(self):
...
+        msg = "Columns must be same length as key"
+        with pytest.raises(ValueError, match=msg):
+            df["a"] = rhs
+
+    def test_setitem_listlike_indexer_duplicate_columns(self):
+        # GH#38604
+        df = DataFrame([[1, 2, 3]], columns=["a", "b", "b"])
+        rhs = DataFrame([[10, 11, 12]], columns=["d", "e", "c"])
+        df[["a", "b"]] = rhs
+        expected = DataFrame([[10, 11, 12]], columns=["a", "b", "b"])
```
Comment (jbrockmendel) [inline pandas/tests/frame/indexing/test_setitem.py:383]: is this obvious?  is rhs.columns irrelevant?

Comment (phofl) [reply pandas/tests/frame/indexing/test_setitem.py:383]: What do you mean with obvious?

Yes columns are irrelevant with setitem

File: pandas/core/frame.py [hunk 4]
```diff
@@ -3223,7 +3223,7 @@ def _setitem_array(self, key, value):
             self._check_setitem_copy()
             self.iloc[indexer] = value
         else:
-            if isinstance(value, DataFrame):
+            if isinstance(value, DataFrame) and self.columns.is_unique:
```
Comment (jbrockmendel) [inline pandas/core/frame.py:3226]: will there be a problem on L3230 if value.columns is not unique?

Comment (phofl) [reply pandas/core/frame.py:3226]: No but maybe at 3227

Comment (phofl) [reply pandas/core/frame.py:3226]: Depends of the key references a dup column

Comment (phofl) [reply pandas/core/frame.py:3226]: let me rephrase this:
We don't get there, becasue 3227 would raise. If we would change the condition to recognize duplicates, we would get a problem in 3229, because the lenght of both lists would not match.
If we also fix this, we have to check that in case of a key which is duplicated in the columns, we select two columns and not only one in 3230 on the rhs

Comment (jbrockmendel) [reply pandas/core/frame.py:3226]: yah the other case im now unsure of is `not all(x in self.columns for x in key)`

Comment (phofl) [reply pandas/core/frame.py:3226]: Yikes, this has ugly side effects. We have to find a way to handle duplicates there instead of dispatching down :(

And improve the test coverage of these cases
